### PR TITLE
ENH: Make numpy floor_divide and remainder agree with Python `//` and `%`.

### PR DESCRIPTION
**Pull request by charris**
This fixes the problems with divmod in #7224 and brings divmod for numpy float64 scalars and arrays into agreement with divmod for Python floats. Tests are added to enforce the new behavior and assures that the results for small integers scaled by powers of two are exact. 

Too keep the scalar and array math in sync, both the scalar math and loop code is based on a new `npy_divmod` function in `npy_math`.

EDIT: I changed the name to `npy_divmod`.

divmod example, Python floats

```
    In [1]: a = 78 * 6e-8

    In [2]: b = 6e-8

    In [3]: divmod(a, b)
    Out[3]: (77.0, 5.999999999999965e-08)
```

Before this commit numpy float64 gave

```
    In [4]: divmod(float64(a), float64(b))
    Out[4]: (78.0, 0.0)
```

After this commit numpy float64 gives

```
    In [4]: divmod(float64(a), float64(b))
    Out[4]: (77.0, 5.9999999999999651e-08)
```

### CODE REVIEW

**Comment by charris** [comment]
Note that pymodf uses fmod, as does Python. That function alone takes 4x longer to run than the old code :( That seems to be the price for a little bit more accuracy.

File: numpy/core/src/npymath/npy_math.c.src [hunk 1]
```diff
@@ -608,6 +608,54 @@ double npy_log2(double x)
...
+    else {
+        /* if mod is zero ensure correct sign */
+        mod = (b > 0) ? 0.0@c@ : -0.0@c@;
+    }
+
+    /* snap quotient to nearest integral value */
+    if (div) {
+        floordiv = npy_floor(div);
+        if (div - floordiv > 0.5)
+            floordiv += 1.0;
```
Comment (pv) [inline numpy/core/src/npymath/npy_math.c.src]: 0.5@c@, 1.0@c@ ? Probably doesn't matter.

File: numpy/core/tests/test_umath.py [hunk 2]
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
...
+        dt = np.typecodes['Float']
+        for dt1, dt2 in itertools.product(dt, dt):
+            for sg1, sg2 in itertools.product((+1, -1), (+1, -1)):
+                fmt = 'dt1: %s, dt2: %s, sg1: %s, sg2: %s'
+                msg = fmt % (dt1, dt2, sg1, sg2)
+                a = np.array(sg1*78*6e-8, dtype=dt1)
+                b = np.array(sg2*6e-8, dtype=dt2)
+                div = np.floor_divide(a, b)
+                rem = np.remainder(a, b)
+                assert_allclose(div*b + rem, a, err_msg=msg)
```
Comment (pv) [inline numpy/core/tests/test_umath.py]: assert_equal? Is this an exact relation?

File: numpy/core/tests/test_scalarmath.py [hunk 3]
```diff
@@ -150,29 +156,87 @@ def test_divmod_basic(self):
                 msg = fmt % (dt1, dt2, sg1, sg2)
                 a = np.array(sg1*71, dtype=dt1)[()]
                 b = np.array(sg2*19, dtype=dt2)[()]
-                div, rem = divmod(a, b)
+                div = self.floordiv(a, b)
+                rem = self.mod(a, b)
                 assert_allclose(div*b + rem, a, err_msg=msg)
```
Comment (pv) [inline numpy/core/tests/test_scalarmath.py]: assert_equal? Is this an exact relation?

File: numpy/core/tests/test_scalarmath.py [hunk 4]
```diff
@@ -150,29 +156,87 @@ def test_divmod_basic(self):
...
         for dt1, dt2 in itertools.product(dt, dt):
             for sg1, sg2 in itertools.product((+1, -1), (+1, -1)):
                 fmt = 'dt1: %s, dt2: %s, sg1: %s, sg2: %s'
                 msg = fmt % (dt1, dt2, sg1, sg2)
                 a = np.array(sg1*78*6e-8, dtype=dt1)[()]
                 b = np.array(sg2*6e-8, dtype=dt2)[()]
-                div, rem = divmod(a, b)
+                div = self.floordiv(a, b)
+                rem = self.mod(a, b)
                 assert_allclose(div*b + rem, a, err_msg=msg)
```
Comment (pv) [inline numpy/core/tests/test_scalarmath.py]: Exact?

File: numpy/core/tests/test_umath.py [hunk 5]
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
...
+                    continue
+                if sg2 == -1 and dt2 in np.typecodes['UnsignedInteger']:
+                    continue
+                fmt = 'dt1: %s, dt2: %s, sg1: %s, sg2: %s'
+                msg = fmt % (dt1, dt2, sg1, sg2)
+                a = np.array(sg1*71, dtype=dt1)
+                b = np.array(sg2*19, dtype=dt2)
+                div = np.floor(a/b)
+                rem = np.remainder(a, b)
+                assert_allclose(div*b + rem, a, err_msg=msg)
```
Comment (pv) [inline numpy/core/tests/test_umath.py]: assert_array_almost_equal_nulp?

File: numpy/core/tests/test_umath.py [hunk 6]
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
...
+            for sg1, sg2 in itertools.product((+1, -1), (+1, -1)):
+                if sg1 == -1 and dt1 in np.typecodes['UnsignedInteger']:
+                    continue
+                if sg2 == -1 and dt2 in np.typecodes['UnsignedInteger']:
+                    continue
+                fmt = 'dt1: %s, dt2: %s, sg1: %s, sg2: %s'
+                msg = fmt % (dt1, dt2, sg1, sg2)
+                a = np.array(sg1*71, dtype=dt1)
+                b = np.array(sg2*19, dtype=dt2)
+                div = np.floor(a/b)
```
Comment (pv) [inline numpy/core/tests/test_umath.py]: floor_divide?

File: numpy/core/tests/test_umath.py [hunk 7]
```diff
@@ -222,6 +223,101 @@ def test_floor_division_complex(self):
...
+        tgtdiv = np.where((tgtdiv == 0.0) & ((b < 0) ^ (a < 0)), -0.0, tgtdiv)
+        tgtrem = np.where((tgtrem == 0.0) & (b < 0), -0.0, tgtrem)
+
+        for dt in np.typecodes['Float']:
+            msg = 'dtype: %s' % (dt,)
+            fa = a.astype(dt)
+            fb = b.astype(dt)
+            div = np.floor(fa/fb)
+            rem = np.remainder(fa, fb)
+            assert_equal(div, tgtdiv, err_msg=msg)
```
Comment (pv) [inline numpy/core/tests/test_umath.py]: floor(a,b) or floor_divide(a,b)?
```
import math
print(math.floor(0.6/0.04), 0.6//0.04, divmod(0.6, 0.04))
#-> 15 14.0 (14.0, 0.039999999999999966)
```

File: numpy/core/src/npymath/npy_math.c.src (hunk 1 above)
Comment (charris) [inline numpy/core/src/npymath/npy_math.c.src]: Fixed.

File: numpy/core/tests/test_umath.py (hunk 2 above)
Comment (charris) [inline numpy/core/tests/test_umath.py]: No, this one is sensitive to roundoff error.

File: numpy/core/tests/test_scalarmath.py (hunk 3 above)
Comment (charris) [inline numpy/core/tests/test_scalarmath.py]: Exact. Hmm, looks redundant with other tests of exactness, should probably be removed.

File: numpy/core/tests/test_scalarmath.py (hunk 4 above)
Comment (charris) [inline numpy/core/tests/test_scalarmath.py]: Not this one.

Comment (pv) [inline numpy/core/tests/test_scalarmath.py]: But then I don't understand this PR. Wasn't the aim to maintain the invariant
``a == (a % b) + (a//b)*b``?

However, I note that this test does pass with `assert_equal`.

**Comment by pv** [comment]
I think it should be clarified what the relation between floor_divide and remainder is, if it is known, and reflect that in the tests in the roundoff cases, in case such invariant is known. In cases where it's known that rounding error can exist, `assert_almost_equal_nulp` should probably be used instead of `assert_allclose`, because the default tolerance of the latter is fairly lax.

File: numpy/core/tests/test_umath.py (hunk 2 above)
Comment (pv) [inline numpy/core/tests/test_umath.py]: The test passes for me also with `assert_equal`.

Comment (charris) [inline numpy/core/tests/test_umath.py]: If `modf` is IEEE compliant, equality should obtain, I just don't trust the libraries that much, especially with old x86 fpus and old libraries. The example itself was the one the exposed the defect in the old implementation. However, I suppose we could use equality just to see what failed...

File: numpy/core/tests/test_scalarmath.py (hunk 4 above)
Comment (charris) [inline numpy/core/tests/test_scalarmath.py]: IEEE compliance requires is that `modf` will be correctly rounded but I don't think we can guarantee that `(a % b) + (a//b)*b` is exact, although it should probably be good to within an ULP

**Comment by charris** [comment]
Updated.

**Pull request merged**